File: network/tools/nodejs/echo_client.js
// 2022-10-01 K.OHWADA
// echo_client.js
// https://kazuhira-r.hatenablog.com/entry/20170912/1505227491

const net = require("net");

const util = require('util');

const sleep = require('sleep');

const args = process.argv.slice(2);


const FORMAT = "test %d";

const SLEEP = 1; // 1 sec

const DATA_LIMIT = 10;

let data_cnt = 0;

let host = "localhost";
let port = 1234;
let message = "Hello";


if (args.length > 2) {
    host = args[0];
    port = parseInt(args[1]);
    message = args[2];
} else if (args.length > 1) {
    // host = "localhost";
    port = parseInt(args[0]);
    message = args[1];
}

const logger = fun => console.log(`[${new Date()}] ${fun.call(null)}`);

const client = net.createConnection({ host: host, port: port }, () => {
    logger(() => "start client");
});

client.on("connect", () => {
    logger(() => `connected server[${host}:${client.remotePort}]`);

// send first message
    const msg = util.format(FORMAT, data_cnt);
    client.write(msg);

    // logger(() => `send message[${message}]`);
});

client.on("data", data => {
    // logger(() => `receive message[${data}]`);
        console.info( `r: ${data}`);

    data_cnt ++;
    if(data_cnt > DATA_LIMIT){
        console.info( `destroy`);
        client.destroy();
        return;
    }

    const msg = util.format(FORMAT, data_cnt);
    sleep.sleep(SLEEP);
    client.write(msg);
});

client.on("close", () => logger(() => "disconnect server"));

File: network/tools/python/echo_client.py
## 2022-10-01 K.OHWADA
# Echo client program
# https://docs.python.org/ja/3.6/library/socket.html#example

import socket
import time

HOST = 'localhost'    # The remote host
PORT = 50007              # The same port as used by the server

RECV_MAX_LEN = 1024
FOR_RANGE = 10
SLEEP = 1 # 1 sec
FORMAT="{:03d}"

with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
    sock.connect((HOST, PORT))
    for i in range(FOR_RANGE):
        msg=FORMAT.format(i)
        sock.sendall( msg.encode() )
        rcv_data = sock.recv(RECV_MAX_LEN)
        data = rcv_data.decode()
        print('r: ', data )
        time.sleep(SLEEP)
print('close')
